examples/cpp/fill-grid-pp2hadron.cpp

```diff
@@ -0,0 +1,245 @@
+        return {s, t, u, x1, x2, 0.0, pt_hadron, y_hadron, 0.0};
+    }
+
+    // sample `z` uniformly between the kinematic limits
+    double z = z_min + (z_max - z_min) * r6;
+    jacobian *= (z_max - z_min);
+
+    return {s, t, u, x1, x2, z, pt_hadron, y_hadron, jacobian};
+}
+
+void fill_grid(pineappl_grid* grid, std::size_t calls) {
+    using std::acosh;
+    using std::fabs;
+    using std::log;
+    using std::sqrt;
+
+    auto rng = std::mt19937();
+    double hbarc2 = 389379372.1;
+
+    // define hadron kinematic ranges
+    double pt_min = 5.0;   // GeV
+    double pt_max = 100.0; // GeV
+    double abs_y_max = 2.4;    // rapidity range
+
+    for (std::size_t i = 0; i != calls; ++i) {
+        auto tmp = pspgen_pp2hadron(rng, 3000.0, 14000.0, pt_min, pt_max, abs_y_max);
+        auto s = tmp.s;
+        auto t = tmp.t;
+        auto u = tmp.u;
+        auto x1 = tmp.x1;
+        auto x2 = tmp.x2;
+        auto z = tmp.z;
+        auto pt_hadron = tmp.pt_hadron;
+        auto y_hadron = tmp.y_hadron;
+        auto jacobian = tmp.jacobian;
+
+        // skip if kinematically forbidden
+        if (jacobian == 0.0 || z <= 0.0) {
+            continue;
+        }
+
+        // apply cuts on hadron kinematics
+        if ((pt_hadron < pt_min) || (pt_hadron > pt_max) || (fabs(y_hadron) > abs_y_max)) {
+            continue;
+        }
+
+        jacobian *= hbarc2 / calls;
+
+        // calculate the partonic cross-section
+        auto weight = jacobian * me_gg2qqbar(s, t, u);
+
+        double q2 = pt_hadron * pt_hadron;
+        std::size_t order = 0;
+        std::size_t channel = 0;
+
+        // define the tuple of kinematic variables `ntuples = (q2, x1, x2, z)`
+        std::vector<double> ntuples = {q2, x1, x2, z};
+
+        // Fill the grid using hadron `pT` as the observable
+        pineappl_grid_fill2(grid, order, pt_hadron, channel, ntuples.data(), weight);
+    }
+}
+
+int main() {
+    // ---
+    // Define the partonic channels and orders that will be filled into the grid
+
+    // specify the number of convolutions: 2 for initial-state PDFs + 1 for FFs
+    std::size_t nb_convolutions = 3;
+
+    // instantiate the channel object
+    auto* channels = pineappl_channels_new(nb_convolutions);
+
+    // specify the contributing channel(s) and the corresponding factor(s)
+    // for the process `gg -> qqbar` we need to sum over the light quarks
+    std::vector<int32_t> pids;
+    std::vector<double> factors;
+    for (int i = -3; i <= 3; ++i) {
+        if (i == 0) continue;
+        pids.insert(pids.end(), {21, 21, i});
+        factors.push_back(1.0);
+    }
+    pineappl_channels_add(channels, pids.size() / nb_convolutions, pids.data(),
+        factors.data());
+
+    // specify the perturbative orders that will be filled into the grid
+    // orders specifies the power of the tuple `orders = (αs, α, lR, lF, lD)`
+    // in this example, we only fill the LO QCD
+    std::vector<uint8_t> orders = {1, 0, 0, 0, 0};
```

Comment:
Below you use C-arrays for other parameters, so you should decide between one of them and `std::vector` IMO. Although there are also `std::array` and `std::unique_ptr<[]>` available. I think in this instance C-arrays are probably the best because you don't allocate.

Reply:
In order to be consistent, and for simplicity because we most of the time need to know the actual size of the objects, I decided to go with `std::vector`.